File: src/main/java/io/metjka/vortex/ui/components/connections/InputAnchor.java
package io.metjka.vortex.ui.components.connections;

import com.google.common.collect.ImmutableMap;
import io.metjka.vortex.ui.BlockContainer;
import io.metjka.vortex.ui.Type;
import io.metjka.vortex.ui.components.blocks.Block;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.value.ObservableValue;
import javafx.collections.ObservableList;
import javafx.fxml.FXML;
import javafx.geometry.Point2D;
import javafx.scene.paint.Color;
import javafx.scene.shape.Shape;

import java.util.Map;
import java.util.Optional;

/**
 * ConnectionAnchor that specifically functions as an input.
 */
public class InputAnchor extends ConnectionAnchor implements Target {

    /**
     * The visual representation of the InputAnchor.
     */
    @FXML
    protected Shape visibleAnchor;

    /**
     * The invisible part of the InputAnchor (the touch zone).
     */
    @FXML
    protected Shape invisibleAnchor;

    /**
     * The thing sticking out of an unconnected InputAnchor.
     */
    @FXML
    protected Shape openWire;

    /**
     * The Optional connection this anchor has.
     */
    private Optional<Connection> connection;

    /**
     * The local type of this anchor
     */
    private Type type;

    /**
     * Property storing the error state.
     */
    private BooleanProperty errorState;

    /**
     * @param block The Block this anchor is connected to.
     */
    public InputAnchor(Block block) {
        super(block);
        this.loadFXML("InputAnchor");

        this.connection = Optional.empty();

        this.errorState = new SimpleBooleanProperty(false);
        this.errorState.addListener(this::checkError);
    }

    /**
     * @param block The Block this anchor is connected to.
     * @param type  The type constraint for this anchor.
     */
    public InputAnchor(Block block, Type type) {
        this(block);
        this.type = type;
    }

    @Override
    public ConnectionAnchor getAssociatedAnchor() {
        return this;
    }

    /**
     * @param state The new error state for this ConnectionAnchor.
     */
    protected void setErrorState(boolean state) {
        this.errorState.set(state);
    }

    /**
     * @return The Optional connection this anchor has.
     */
    public Optional<Connection> getConnection() {
        return this.connection;
    }

    /**
     * Sets the connection of this anchor.
     *
     * @param connection The connection to set.
     */
    protected void setConnection(Connection connection) {
        this.connection = Optional.of(connection);
        this.openWire.setVisible(false);
    }

    @Override
    public void removeConnections() {
        if (this.connection.isPresent()) {
            Connection conn = this.connection.get();
            this.connection = Optional.empty();
            conn.remove();
        }
        this.setErrorState(false);
        this.openWire.setVisible(true);
        if (this.getWireInProgress() != null) {
            this.getWireInProgress().remove();
        }
    }

    @Override
    public boolean hasConnection() {
        return this.connection.isPresent();
    }

    /**
     * @return True if this anchor has an error free connection
     */
    public boolean hasValidConnection() {
        return this.connection.isPresent() && !(this.errorState.get() || this.connection.get().hasScopeError());
    }

    @Override
    public Point2D getAttachmentPoint() {
        return this.getPane().sceneToLocal(this.localToScene(new Point2D(0, -7)));
    }

    /**
     * @return the local type of this anchor
     */
    public Type getType() {
        return this.type;
    }

    /**
     * @return the string representation of the in- or output type.
     */
    public final String getStringType() {
        return type.toString();
    }

    public void setExactRequiredType(Type type) {
        this.type = type;
    }

    /**
     * @return Optional of the connection's opposite output anchor.
     */
    public Optional<OutputAnchor> getOppositeAnchor() {
        return this.connection.map(c -> c.getStartAnchor());
    }

    /**
     * ChangeListener that will set the error state if isConnected().
     */
    public void checkError(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) {
        ObservableList<String> style = this.visibleAnchor.getStyleClass();
        style.removeAll("error");
        if (newValue) {
            style.add("error");
        }
    }

    @Override
    protected void setNearbyWireReaction(int goodness) {

        switch (goodness) {
            case DrawWire.GOOD_TYPE_REACTION: {
                this.openWire.setStroke(Color.DARKGREEN);
                this.openWire.setStrokeWidth(5);
                this.visibleAnchor.setFill(Color.DARKGREEN);
                break;
            }
            case DrawWire.NEUTRAL_TYPE_REACTION: {
                this.openWire.setStroke(Color.DODGERBLUE);
                this.openWire.setStrokeWidth(5);
                this.visibleAnchor.setFill(Color.DODGERBLUE);
                break;
            }
            case DrawWire.WRONG_TYPE_REACTION: {
                this.openWire.setStroke(Color.RED);
                this.openWire.setStrokeWidth(3);
                this.visibleAnchor.setFill(Color.RED);
                break;
            }
            default: {
                this.openWire.setStroke(Color.BLACK);
                this.openWire.setStrokeWidth(3);
                this.visibleAnchor.setFill(Color.BLACK);
                break;
            }
        }
    }

    @Override
    public void setWireInProgress(DrawWire wire) {
        super.setWireInProgress(wire);
        if (wire == null) {
            this.openWire.setVisible(!this.hasConnection());
            this.invisibleAnchor.setMouseTransparent(false);
        } else {
            this.openWire.setVisible(false);
            this.invisibleAnchor.setMouseTransparent(true);
        }
    }

    /**
     * Called when the VisualState changed.
     */
    public void invalidateVisualState() {
        this.connection.ifPresent(connection -> connection.invalidateVisualState());
    }

    @Override
    public BlockContainer getContainer() {
        return this.block.getContainer();
    }

    @Override
    public String toString() {
        return "InputAnchor for " + this.block;
    }

    @Override
    public Map<String, Object> toBundle() {
        ImmutableMap.Builder<String, Object> bundle = ImmutableMap.builder();
        bundle.put(BLOCK_LABEL, this.block.hashCode());
        bundle.put(ANCHOR_LABEL, block.getAllInputs().indexOf(this));
        return bundle.build();
    }

}

File: src/main/java/io/metjka/vortex/ui/connections/InputAnchor.kt
package io.metjka.vortex.ui.connections

import io.metjka.vortex.ui.BlockContainer
import io.metjka.vortex.ui.Type
import io.metjka.vortex.ui.blocks.Block
import javafx.beans.property.BooleanProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.value.ObservableValue
import javafx.fxml.FXML
import javafx.geometry.Point2D
import javafx.scene.paint.Color
import javafx.scene.shape.Shape
import java.util.*

class InputAnchor<T>(block: Block, val type: Type) : ConnectionAnchor(block), Target {

    @FXML
    var visibleAnchor: Shape? = null

    @FXML
    var invisibleAnchor: Shape? = null

    @FXML
    var openWire: Shape? = null

    private var connectionDep: Optional<ConnectionDep<T>>

    var errorState: BooleanProperty

    init {
        loadFXML(this::class.simpleName)

        connectionDep = Optional.empty<ConnectionDep<T>>()
        errorState = SimpleBooleanProperty(false)
        errorState.addListener { observable, oldValue, newValue -> checkError(observable, oldValue, newValue) }
    }

    fun checkError(observable: ObservableValue<out Boolean>, oldValue: Boolean?, newValue: Boolean?) {
        val style = this.visibleAnchor?.styleClass
        style?.removeAll("error")
        if (newValue!!) {
            style?.add("error")
        }
    }

    override fun getAssociatedAnchor(): ConnectionAnchor {
        return this
    }

    fun getConnection(): Optional<ConnectionDep<T>> {
        return connectionDep
    }

    fun getOppositeAnchor(): Optional<OutputAnchor<T>> {
        return this.connectionDep.map { c -> c.start }
    }

    fun setConnection(connectionDep: ConnectionDep<T>) {
        this.connectionDep = Optional.of(connectionDep)
        this.openWire?.isVisible = false
    }

    override fun removeConnections() {
        if (connectionDep.isPresent) {
            val get = connectionDep.get()
            connectionDep = Optional.empty()
            get.remove()
        }
        errorState.value = false
        openWire?.isVisible = true

        if (wireInProgress != null) {
            wireInProgress?.remove()
        }

    }

    fun onUpdate() {
        block.sendUpdateDownStream()
    }

    override fun hasConnection(): Boolean {
        return connectionDep.isPresent
    }

    override fun getAttachmentPoint(): Point2D {
        return topLevelPane.sceneToLocal(localToScene(Point2D(0.0, -7.0)))
    }

    override fun setNearbyWireReaction(goodness: Int) {

        when (goodness) {
            DrawWire.GOOD_TYPE_REACTION -> {
                this.openWire?.stroke = Color.DARKGREEN
                this.openWire?.strokeWidth = 5.0
                this.visibleAnchor?.fill = Color.DARKGREEN
            }
            DrawWire.NEUTRAL_TYPE_REACTION -> {
                this.openWire?.stroke = Color.DODGERBLUE
                this.openWire?.strokeWidth = 5.0
                this.visibleAnchor?.fill = Color.DODGERBLUE
            }
            DrawWire.WRONG_TYPE_REACTION -> {
                this.openWire?.stroke = Color.RED
                this.openWire?.strokeWidth = 3.0
                this.visibleAnchor?.fill = Color.RED
            }
            else -> {
                this.openWire?.stroke = Color.BLACK
                this.openWire?.strokeWidth = 3.0
                this.visibleAnchor?.fill = Color.BLACK
            }
        }
    }

    override fun getContainer(): BlockContainer {
        return block.getContainer()
    }

    fun invalidateVisualState() {
        this.connectionDep.ifPresent { it.invalidateVisualState() }
    }

    override fun toString(): String {
        return "InputAnchor for $block"
    }

    override fun toBundle(): Map<String, Any> {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}
